test/unit/lib/classes/Variables.test.js

```diff
@@ -2741,6 +2741,162 @@ module.exports = {
       expect(logWarningSpy.args[0][0]).to.contain('file');
     });
   });
+  describe('#assertFound()', () => {
```

Comment:
I think it would be better to try to use `runServerless`-based tests to test out the newly introduced funtionality here. We try to migrate all tests to `runServerless`, here you can see an example on how that might look like: https://github.com/G-Rath/serverless/blob/master/test/unit/lib/classes/Variables.test.js#L2745-L2783

Reply:
Sure, I'll give it a try - however, if I understand things correctly that'll result in a lot of fixtures (and possible high slowdown since those tests are expensive?) since we error on the first unresolved variable.

So if we want to have the same test coverage as before we'd need to have a test/fixture for each possible value of `varType`, for each possible mode supported by `unresolvedVariablesNotificationMode`.

A possible way around this could be keep the `_describeVariableType` method and unit test that, and use `runServerless` tests for the rest (which'd mean only needing 2 or 3 tests)

Reply:
> Sure, I'll give it a try - however, if I understand things correctly that'll result in a lot of fixtures 

We should reuse existing fixture (`variables`), which we can simply extend for every case

> (and possible high slowdown since those tests are expensive?) 

They're somewhat expensive, and it's good to minimize `runServerless` runs when possible, but all error cases are test ed individually, and in this case we need to do different `runServerless` run per fixture.

Test best if are added at the bottom of tests files, where we already have some `runServerless` based tests added

Reply:
So I've been trying to write a single test for now, but no matter what I do `unresolvedVariablesNotificationMode` doesn't seem to be making it into the config - is there anything I might have missed?

Reply:
@G-Rath can you commit the code, that you feel that should work and doesn't work for you?

Reply:
Sure - I've pushed up my changes as-is.

I've tried both with the fixture and setting the value via `configExt`.

<details>
<summary>This is what my test runner logs</summary>

```
C:\Windows\system32\wsl.exe --distribution Ubuntu --exec /bin/sh -c "export PATH=\"/home/g-rath/.nodenv/shims:$PATH\" && cd /c/Users/G-Rath/workspace/projects-oss/serverless && /home/g-rath/.nodenv/shims/node /c/Users/G-Rath/workspace/projects-oss/serverless/node_modules/mocha/bin/mocha --ui bdd --reporter '/c/Program Files/JetBrains/IntelliJ IDEA 2020.3/plugins/NodeJS/js/mocha-intellij/lib/mochaIntellijReporter.js' /c/Users/G-Rath/workspace/projects-oss/serverless/test/unit/lib/classes/Variables.test.js --grep '^Variables variable resolving when unresolvedVariablesNotificationMode is set to \"error\" should error for missing env type variables$'"
{
  serverless: <ref *1> Serverless {
    providers: { aws: [AwsProvider] },
    version: '2.17.0',
    yamlParser: YamlParser { serverless: [Circular *1] },
    utils: Utils { serverless: [Circular *1] },
    service: Service {
      serverless: [Circular *1],
      service: 'test-variables-unresolved-error-9ogb4he',
      serviceObject: [Object],
      provider: [Object],
      custom: [Object],
      plugins: undefined,
      pluginsData: {},
      functions: {},
      resources: undefined,
      package: {},
      configValidationMode: 'warn',
      disabledDeprecations: undefined,
      serviceFilename: 'serverless.yml',
      app: undefined,
      org: undefined,
      layers: {},
      outputs: undefined,
      initialServerlessConfig: [Object],
      isDashboardMonitoringPreconfigured: false
    },
    variables: Variables {
      serverless: [Circular *1],
      service: [Object],
      tracker: [PromiseTracker],
      deep: [],
      deepRefSyntax: /(\${)?deep:\d+(\.[^}]+)*()}?/,
      overwriteSyntax: /\s*(?:,\s*)+/g,
      fileRefSyntax: /^file\(([^?%*:|"<>]+?)\)/g,
      slsRefSyntax: /^sls:/g,
      envRefSyntax: /^env:/g,
      optRefSyntax: /^opt:/g,
      selfRefSyntax: /^self:/g,
      stringRefSyntax: /(?:^('|").*?\1$)/g,
      boolRefSyntax: /(?:^(true|false)$)/g,
      intRefSyntax: /(?:^\d+$)/g,
      s3RefSyntax: /^(?:\${)?s3:(.+?)\/(.+)$/,
      cfRefSyntax: /^(?:\${)?cf(?:\.([a-zA-Z0-9-]+))?:(.+?)\.(.+)$/,
      ssmRefSyntax: /^(?:\${)?ssm(?:\.([a-zA-Z0-9-]+))?:([a-zA-Z0-9_.\-/]+)[~]?(true|false|split)?/,
      strToBoolRefSyntax: /^(?:\${)?strToBool\(([a-zA-Z0-9_.\-/]+)\)/,
      variableResolvers: [Array],
      options: [Object],
      variableSyntax: /\${([^{}:]+?(?:\(|:)(?:[^:{}][^{}]*?)?)}/g
    },
    pluginManager: PluginManager {
      serverless: [Circular *1],
      serverlessConfigFile: [Object],
      cliOptions: [Object],
      cliCommands: [Array],
      pluginIndependentCommands: [Set],
      plugins: [Array],
      externalPlugins: Set(0) {},
      commands: [Object],
      aliases: {},
      hooks: [Object],
      deprecatedEvents: [Object]
    },
    configSchemaHandler: ConfigSchemaHandler { serverless: [Circular *1], schema: [Object] },
    cliInputArgv: [ 'print' ],
    config: Config {
      serverless: [Circular *1],
      serverlessPath: '/c/Users/G-Rath/workspace/projects-oss/serverless/lib',
      servicePath: '/tmp/tmpdirs-serverless/1039/1ecc6e'
    },
    classes: {
      CLI: [Function: CLI],
      YamlParser: [Function: YamlParser],
      Utils: [Function: Utils],
      Service: [Function: Service],
      Variables: [Function: Variables],
      Error: [Function: ServerlessError],
      PluginManager: [Function: PluginManager],
      ConfigSchemaHandler: [Function: ConfigSchemaHandler]
    },
    serverlessDirPath: '/tmp/tmpdirs-serverless/1039/84b7b1/.serverless',
    isStandaloneExecutable: false,
    isLocallyInstalled: false,
    isInvokedByGlobalInstallation: false,
    triggeredDeprecations: Set(2) {
      'VARIABLES_ERROR_ON_UNRESOLVED',
      'LAMBDA_HASHING_VERSION_V2'
    },
    instanceId: '1609879725108',
    cli: CLI {
      serverless: [Circular *1],
      inputArray: [Array],
      loadedPlugins: [Array],
      loadedCommands: [Object],
      log: [Function (anonymous)]
    },
    processedInput: { commands: [Array], options: [Object] },
    httpApiEventsPlugin: HttpApiEvents {
      serverless: [Circular *1],
      provider: [AwsProvider],
      hooks: [Object]
    }
  },
  stdoutData: 'undefined\n' +
    'Serverless: \x1B[38;5;214mDeprecation warning: Unresolved variable references will be communicated with a thrown error, starting from next major\x1B[39m\n' +
    '            \x1B[38;5;214mMore Info: https://www.serverless.com/framework/docs/deprecations/#VARIABLES_ERROR_ON_UNRESOLVED\x1B[39m\n' +
    ' \n' +
    '\x1B[33m Serverless Warning --------------------------------------\x1B[39m\n' +
    ' \n' +
    "  A valid environment variable to satisfy the declaration 'env:value' could not be found.\n" +
    ' \n' +
    'Serverless: \x1B[38;5;214mDeprecation warning: Starting with next major version, default value of provider.lambdaHashingVersion will be equal to "20201221"\x1B[39m\n' +
    '            \x1B[38;5;214mMore Info: https://www.serverless.com/framework/docs/deprecations/#LAMBDA_HASHING_VERSION_V2\x1B[39m\n' +
    'service:\n' +
    '  name: test-variables-unresolved-error-9ogb4he\n' +
    'unresolvedVariablesNotificationMode: error\n' +
    'provider:\n' +
    '  stage: dev\n' +
    '  region: us-east-1\n' +
    '  name: aws\n' +
    '  runtime: nodejs12.x\n' +
    '  versionFunctions: true\n' +
    "  variableSyntax: '\\${([^{}:]+?(?:\\(|:)(?:[^:{}][^{}]*?)?)}'\n" +
    'custom:\n' +
    '  hello: world\n' +
    '\n',
  cfTemplate: undefined,
  awsNaming: <ref *2> {
    provider: <ref *3> AwsProvider {
      naming: [Circular *2],
      options: [Object],
      provider: [Circular *3],
      serverless: [Serverless],
      sdk: [Object],
      hooks: [Object],
      requestCache: {},
      requestQueue: [Queue],
      cachedCredentials: null
    },
    normalizeName: [Function: normalizeName],
    normalizeNameToAlphaNumericOnly: [Function: normalizeNameToAlphaNumericOnly],
    toStartCase: [Function: toStartCase],
    normalizePathPart: [Function: normalizePathPart],
    getServiceEndpointRegex: [Function: getServiceEndpointRegex],
    getStackName: [Function: getStackName],
    getServiceArtifactName: [Function: getServiceArtifactName],
    getFunctionArtifactName: [Function: getFunctionArtifactName],
    getLayerArtifactName: [Function: getLayerArtifactName],
    getServiceStateFileName: [Function: getServiceStateFileName],
    getCompiledTemplateFileName: [Function: getCompiledTemplateFileName],
    getCompiledTemplateS3Suffix: [Function: getCompiledTemplateS3Suffix],
    getCoreTemplateFileName: [Function: getCoreTemplateFileName],
    getRolePath: [Function: getRolePath],
    getRoleName: [Function: getRoleName],
    getRoleLogicalId: [Function: getRoleLogicalId],
    getPolicyName: [Function: getPolicyName],
    getLogGroupLogicalId: [Function: getLogGroupLogicalId],
    getLogGroupName: [Function: getLogGroupName],
    getNormalizedFunctionName: [Function: getNormalizedFunctionName],
    extractLambdaNameFromArn: [Function: extractLambdaNameFromArn],
    extractAuthorizerNameFromArn: [Function: extractAuthorizerNameFromArn],
    getLambdaLogicalId: [Function: getLambdaLogicalId],
    getLambdaEventConfigLogicalId: [Function: getLambdaEventConfigLogicalId],
    getLambdaLayerLogicalId: [Function: getLambdaLayerLogicalId],
    getLambdaLayerPermissionLogicalId: [Function: getLambdaLayerPermissionLogicalId],
    getLambdaLogicalIdRegex: [Function: getLambdaLogicalIdRegex],
    getLambdaVersionLogicalId: [Function: getLambdaVersionLogicalId],
    getCodeDeployApplicationLogicalId: [Function: getCodeDeployApplicationLogicalId],
    getCodeDeployDeploymentGroupLogicalId: [Function: getCodeDeployDeploymentGroupLogicalId],
    getCodeDeployRoleLogicalId: [Function: getCodeDeployRoleLogicalId],
    getLambdaProvisionedConcurrencyAliasLogicalId: [Function: getLambdaProvisionedConcurrencyAliasLogicalId],
    getLambdaProvisionedConcurrencyAliasName: [Function: getLambdaProvisionedConcurrencyAliasName],
    getLambdaVersionOutputLogicalId: [Function: getLambdaVersionOutputLogicalId],
    getLambdaLayerOutputLogicalId: [Function: getLambdaLayerOutputLogicalId],
    getLambdaLayerHashOutputLogicalId: [Function: getLambdaLayerHashOutputLogicalId],
    getLambdaLayerS3KeyOutputLogicalId: [Function: getLambdaLayerS3KeyOutputLogicalId],
    getWebsocketsApiName: [Function: getWebsocketsApiName],
    getWebsocketsApiLogicalId: [Function: getWebsocketsApiLogicalId],
    getWebsocketsIntegrationLogicalId: [Function: getWebsocketsIntegrationLogicalId],
    getLambdaWebsocketsPermissionLogicalId: [Function: getLambdaWebsocketsPermissionLogicalId],
    getNormalizedWebsocketsRouteKey: [Function: getNormalizedWebsocketsRouteKey],
    getWebsocketsRouteLogicalId: [Function: getWebsocketsRouteLogicalId],
    getWebsocketsDeploymentLogicalId: [Function: getWebsocketsDeploymentLogicalId],
    getWebsocketsRouteResponseLogicalId: [Function: getWebsocketsRouteResponseLogicalId],
    getWebsocketsStageLogicalId: [Function: getWebsocketsStageLogicalId],
    getWebsocketsAuthorizerLogicalId: [Function: getWebsocketsAuthorizerLogicalId],
    getWebsocketsLogGroupLogicalId: [Function: getWebsocketsLogGroupLogicalId],
    getApiGatewayName: [Function: getApiGatewayName],
    generateApiGatewayDeploymentLogicalId: [Function: generateApiGatewayDeploymentLogicalId],
    getRestApiLogicalId: [Function: getRestApiLogicalId],
    getNormalizedAuthorizerName: [Function: getNormalizedAuthorizerName],
    getAuthorizerLogicalId: [Function: getAuthorizerLogicalId],
    normalizePath: [Function: normalizePath],
    getResourceLogicalId: [Function: getResourceLogicalId],
    extractResourceId: [Function: extractResourceId],
    normalizeMethodName: [Function: normalizeMethodName],
    getMethodLogicalId: [Function: getMethodLogicalId],
    getValidatorLogicalId: [Function: getValidatorLogicalId],
    getModelLogicalId: [Function: getModelLogicalId],
    getApiKeyLogicalId: [Function: getApiKeyLogicalId],
    getApiKeyLogicalIdRegex: [Function: getApiKeyLogicalIdRegex],
    getUsagePlanLogicalId: [Function: getUsagePlanLogicalId],
    getUsagePlanKeyLogicalId: [Function: getUsagePlanKeyLogicalId],
    getStageLogicalId: [Function: getStageLogicalId],
    getApiGatewayLogGroupLogicalId: [Function: getApiGatewayLogGroupLogicalId],
    getDeploymentBucketLogicalId: [Function: getDeploymentBucketLogicalId],
    getDeploymentBucketOutputLogicalId: [Function: getDeploymentBucketOutputLogicalId],
    getDeploymentBucketPolicyLogicalId: [Function: getDeploymentBucketPolicyLogicalId],
    normalizeBucketName: [Function: normalizeBucketName],
    getBucketLogicalId: [Function: getBucketLogicalId],
    normalizeTopicName: [Function: normalizeTopicName],
    getTopicLogicalId: [Function: getTopicLogicalId],
    getTopicDLQPolicyLogicalId: [Function: getTopicDLQPolicyLogicalId],
    getScheduleId: [Function: getScheduleId],
    getScheduleLogicalId: [Function: getScheduleLogicalId],
    getStreamLogicalId: [Function: getStreamLogicalId],
    getStreamConsumerName: [Function: getStreamConsumerName],
    getStreamConsumerLogicalId: [Function: getStreamConsumerLogicalId],
    getIotLogicalId: [Function: getIotLogicalId],
    getLambdaIotPermissionLogicalId: [Function: getLambdaIotPermissionLogicalId],
    getIotFleetProvisioningLogicalId: [Function: getIotFleetProvisioningLogicalId],
    getLambdaIotFleetProvisioningPermissionLogicalId: [Function: getLambdaIotFleetProvisioningPermissionLogicalId],
    getCloudWatchEventId: [Function: getCloudWatchEventId],
    getCloudWatchEventLogicalId: [Function: getCloudWatchEventLogicalId],
    getCloudWatchLogLogicalId: [Function: getCloudWatchLogLogicalId],
    getCognitoUserPoolLogicalId: [Function: getCognitoUserPoolLogicalId],
    getQueueLogicalId: [Function: getQueueLogicalId],
    getMSKEventLogicalId: [Function: getMSKEventLogicalId],
    getAlbTargetGroupLogicalId: [Function: getAlbTargetGroupLogicalId],
    getAlbTargetGroupNameTagValue: [Function: getAlbTargetGroupNameTagValue],
    getAlbTargetGroupName: [Function: getAlbTargetGroupName],
    getAlbListenerRuleLogicalId: [Function: getAlbListenerRuleLogicalId],
    getLambdaS3PermissionLogicalId: [Function: getLambdaS3PermissionLogicalId],
    getLambdaSnsPermissionLogicalId: [Function: getLambdaSnsPermissionLogicalId],
    getLambdaSnsSubscriptionLogicalId: [Function: getLambdaSnsSubscriptionLogicalId],
    getLambdaSchedulePermissionLogicalId: [Function: getLambdaSchedulePermissionLogicalId],
    getLambdaCloudWatchEventPermissionLogicalId: [Function: getLambdaCloudWatchEventPermissionLogicalId],
    getLambdaApiGatewayPermissionLogicalId: [Function: getLambdaApiGatewayPermissionLogicalId],
    getLambdaHttpApiPermissionLogicalId: [Function: getLambdaHttpApiPermissionLogicalId],
    getLambdaAlexaSkillPermissionLogicalId: [Function: getLambdaAlexaSkillPermissionLogicalId],
    getLambdaAlexaSmartHomePermissionLogicalId: [Function: getLambdaAlexaSmartHomePermissionLogicalId],
    getLambdaCloudWatchLogPermissionLogicalId: [Function: getLambdaCloudWatchLogPermissionLogicalId],
    getLambdaCognitoUserPoolPermissionLogicalId: [Function: getLambdaCognitoUserPoolPermissionLogicalId],
    getLambdaAlbPermissionLogicalId: [Function: getLambdaAlbPermissionLogicalId],
    getLambdaRegisterTargetPermissionLogicalId: [Function: getLambdaRegisterTargetPermissionLogicalId],
    getCustomResourcesArtifactName: [Function: getCustomResourcesArtifactName],
    getCustomResourcesRoleLogicalId: [Function: getCustomResourcesRoleLogicalId],
    getCustomResourceS3HandlerFunctionName: [Function: getCustomResourceS3HandlerFunctionName],
    getCustomResourceS3HandlerFunctionLogicalId: [Function: getCustomResourceS3HandlerFunctionLogicalId],
    getCustomResourceS3ResourceLogicalId: [Function: getCustomResourceS3ResourceLogicalId],
    getCustomResourceCognitoUserPoolHandlerFunctionName: [Function: getCustomResourceCognitoUserPoolHandlerFunctionName],
    getCustomResourceCognitoUserPoolHandlerFunctionLogicalId: [Function: getCustomResourceCognitoUserPoolHandlerFunctionLogicalId],
    getCustomResourceCognitoUserPoolResourceLogicalId: [Function: getCustomResourceCognitoUserPoolResourceLogicalId],
    getCustomResourceEventBridgeHandlerFunctionName: [Function: getCustomResourceEventBridgeHandlerFunctionName],
    getCustomResourceEventBridgeHandlerFunctionLogicalId: [Function: getCustomResourceEventBridgeHandlerFunctionLogicalId],
    getCustomResourceEventBridgeResourceLogicalId: [Function: getCustomResourceEventBridgeResourceLogicalId],
    getCustomResourceApiGatewayAccountCloudWatchRoleHandlerFunctionName: [Function: getCustomResourceApiGatewayAccountCloudWatchRoleHandlerFunctionName],
    getCustomResourceApiGatewayAccountCloudWatchRoleHandlerFunctionLogicalId: [Function: getCustomResourceApiGatewayAccountCloudWatchRoleHandlerFunctionLogicalId],
    getCustomResourceApiGatewayAccountCloudWatchRoleResourceLogicalId: [Function: getCustomResourceApiGatewayAccountCloudWatchRoleResourceLogicalId],
    getCloudFrontDistributionLogicalId: [Function: getCloudFrontDistributionLogicalId],
    getCloudFrontDistributionDomainNameLogicalId: [Function: getCloudFrontDistributionDomainNameLogicalId],
    getLambdaAtEdgeInvokePermissionLogicalId: [Function: getLambdaAtEdgeInvokePermissionLogicalId],
    getCloudFrontOriginId: [Function: getCloudFrontOriginId],
    getCloudFrontCachePolicyLogicalId: [Function: getCloudFrontCachePolicyLogicalId],
    getHttpApiName: [Function: getHttpApiName],
    getHttpApiLogicalId: [Function: getHttpApiLogicalId],
    getHttpApiStageLogicalId: [Function: getHttpApiStageLogicalId],
    getHttpApiIntegrationLogicalId: [Function: getHttpApiIntegrationLogicalId],
    getHttpApiRouteLogicalId: [Function: getHttpApiRouteLogicalId],
    getHttpApiAuthorizerLogicalId: [Function: getHttpApiAuthorizerLogicalId],
    getHttpApiLogGroupLogicalId: [Function: getHttpApiLogGroupLogicalId],
    getHttpApiLogGroupName: [Function: getHttpApiLogGroupName]
  },
  fixtureData: {
    servicePath: '/tmp/tmpdirs-serverless/1039/1ecc6e',
    serviceConfig: {
      service: 'test-variables-unresolved-error-9ogb4he',
      unresolvedVariablesNotificationMode: 'error',
      provider: [Object],
      custom: [Object]
    },
    updateConfig: [Function: updateConfig]
  }
}
```
</details>

Reply:
I don't see any working tests committed in. Can you pass a link to `runServerless` configuration that doesn't work for you?

Reply:
https://github.com/serverless/serverless/pull/8710/files#diff-4a96776f1029649178f60a5029598be51f460a87df42bd4db2c19598e7c2f7b8R2778-R2813

Reply:
Thanks, technically it works as expected, it's just that top level config properties, currently require additional manual handling here: https://github.com/serverless/serverless/blob/66aa66fbfe363edeb4123d709890a7c78f74b571/lib/classes/Service.js#L120-L174

(sorry, I know that it's by no means easily discoverable)

Also let's not create a new fixture, but reuse `variables` fixture and extend it via `configExt`

Reply:
> (and actually sometimes the tests just run out of memory :/)

It may mean that for some reason configuration triggers `custom` resource installation (it involves `npm install` behind the scene).  Still it'll be good to diagnose why given configuration triggers it (variables testing I think should not involve generation oif custom resources).

Reply:
Thanks, but that doesn't seem to have had any effect.

I've cleaned up my tests so that they now represent tests I'd expect to be passing, but are not.

Sticking `console.log(saveYaml(configObject));` in `setup-fixture-engines.js` gives this:

```
service: test-variables-9omgob6
provider:
  name: aws
  runtime: nodejs12.x
custom:
  importedFile: '${file(config.json)}'
  importedFileWithKey: '${file(config.json):foo}'
  awsVariable: '${AWS::Region}'
  cloudFormationReference: '${AnotherResource}'
  selfReference: '${self:custom.importedFileWithKey}'
  serviceReference: '${self:}'
  nestedRef: nestedVal
  nestedVal:
    prop: resolvedNested
  nestedReference: '${self:custom.${self:custom.nestedRef}.prop}'
  myVariable: '${env:value}'
unresolvedVariablesNotificationMode: error
```

<details>
<summary>
But it seems this is removed later, as it's not present in the `print` output
</summary>

```
2021-01-05T22:36:40.651Z ● mocha 
2021-01-05T22:36:40.653Z ● mocha Variables
2021-01-05T22:36:40.658Z ● mocha Variables > variable resolving
2021-01-05T22:36:40.659Z ● mocha Variables > variable resolving > when unresolvedVariablesNotificationMode is set to "error"
2021-01-05T22:36:40.659Z ● mocha Variables > variable resolving > when unresolvedVariablesNotificationMode is set to "error" > should error for missing env type variables
2021-01-05T22:36:40.681Z ℹ serverless:test setup variables fixture at /tmp/tmpdirs-serverless/45eb/85c788
2021-01-05T22:36:41.848Z ℹ serverless:test:stdout Serverless: Deprecation warning: Unresolved variable references will be communicated with a thrown error, starting from next major

2021-01-05T22:36:41.848Z ℹ serverless:test:stdout             More Info: https://www.serverless.com/framework/docs/deprecations/#VARIABLES_ERROR_ON_UNRESOLVED

2021-01-05T22:36:41.848Z ℹ serverless:test:stdout  

2021-01-05T22:36:41.848Z ℹ serverless:test:stdout  Serverless Warning --------------------------------------

2021-01-05T22:36:41.848Z ℹ serverless:test:stdout  

2021-01-05T22:36:41.848Z ℹ serverless:test:stdout   A valid environment variable to satisfy the declaration 'env:value' could not be found.

2021-01-05T22:36:41.848Z ℹ serverless:test:stdout  

2021-01-05T22:36:43.016Z ℹ serverless:test:stdout Serverless: Deprecation warning: Starting with next major version, default value of provider.lambdaHashingVersion will be equal to "20201221"

2021-01-05T22:36:43.016Z ℹ serverless:test:stdout             More Info: https://www.serverless.com/framework/docs/deprecations/#LAMBDA_HASHING_VERSION_V2

2021-01-05T22:36:43.118Z ℹ serverless:test:stdout service:
  name: test-variables-9omgob6
provider:
  stage: dev
  region: us-east-1
  name: aws
  runtime: nodejs12.x
  versionFunctions: true
  variableSyntax: '\${([^{}:]+?(?:\(|:)(?:[^:{}][^{}]*?)?)}'
custom:
  importedFile:
    foo: bar
  importedFileWithKey: bar
  awsVariable: '${AWS::Region}'
  cloudFormationReference: '${AnotherResource}'
  selfReference: bar
  serviceReference:
    serverless:
      providers:
        aws:
          naming:
            provider:
              $ref: >-
                $["custom"]["serviceReference"]["serverless"]["providers"]["aws"]
          options: {}
          provider:
            $ref: '$["custom"]["serviceReference"]["serverless"]["providers"]["aws"]'
          serverless:
            $ref: '$["custom"]["serviceReference"]["serverless"]'
          sdk:
            util:
              environment: nodejs
              base64: {}
              buffer: {}
              string: {}
              ini: {}
              fn: {}
              date: {}
              crypto:
                crc32Table:
                  - 0
                  - 1996959894
                  - 3993919788
                  - 2567524794
                  - 124634137
                  - 1886057615
                  - 3915621685
                  - 2657392035
                  - 249268274
                  - 2044508324
                  - 3772115230
                  - 2547177864
                  - 162941995
                  - 2125561021
                  - 3887607047
                  - 2428444049
                  - 498536548
                  - 1789927666
                  - 4089016648
                  - 2227061214
                  - 450548861
                  - 1843258603
                  - 4107580753
                  - 2211677639
                  - 325883990
                  - 1684777152
                  - 4251122042
                  - 2321926636
                  - 335633487
                  - 1661365465
                  - 4195302755
                  - 2366115317
                  - 997073096
                  - 1281953886
                  - 3579855332
                  - 2724688242
                  - 1006888145
                  - 1258607687
                  - 3524101629
                  - 2768942443
                  - 901097722
                  - 1119000684
                  - 3686517206
                  - 2898065728
                  - 853044451
                  - 1172266101
                  - 3705015759
                  - 2882616665
                  - 651767980
                  - 1373503546
                  - 3369554304
                  - 3218104598
                  - 565507253
                  - 1454621731
                  - 3485111705
                  - 3099436303
                  - 671266974
                  - 1594198024
                  - 3322730930
                  - 2970347812
                  - 795835527
                  - 1483230225
                  - 3244367275
                  - 3060149565
                  - 1994146192
                  - 31158534
                  - 2563907772
                  - 4023717930
                  - 1907459465
                  - 112637215
                  - 2680153253
                  - 3904427059
                  - 2013776290
                  - 251722036
                  - 2517215374
                  - 3775830040
                  - 2137656763
                  - 141376813
                  - 2439277719
                  - 3865271297
                  - 1802195444
                  - 476864866
                  - 2238001368
                  - 4066508878
                  - 1812370925
                  - 453092731
                  - 2181625025
                  - 4111451223
                  - 1706088902
                  - 314042704
                  - 2344532202
                  - 4240017532
                  - 1658658271
                  - 366619977
                  - 2362670323
                  - 4224994405
                  - 1303535960
                  - 984961486
                  - 2747007092
                  - 3569037538
                  - 1256170817
                  - 1037604311
                  - 2765210733
                  - 3554079995
                  - 1131014506
                  - 879679996
                  - 2909243462
                  - 3663771856
                  - 1141124467
                  - 855842277
                  - 2852801631
                  - 3708648649
                  - 1342533948
                  - 654459306
                  - 3188396048
                  - 3373015174
                  - 1466479909
                  - 544179635
                  - 3110523913
                  - 3462522015
                  - 1591671054
                  - 702138776
                  - 2966460450
                  - 3352799412
                  - 1504918807
                  - 783551873
                  - 3082640443
                  - 3233442989
                  - 3988292384
                  - 2596254646
                  - 62317068
                  - 1957810842
                  - 3939845945
                  - 2647816111
                  - 81470997
                  - 1943803523
                  - 3814918930
                  - 2489596804
                  - 225274430
                  - 2053790376
                  - 3826175755
                  - 2466906013
                  - 167816743
                  - 2097651377
                  - 4027552580
                  - 2265490386
                  - 503444072
                  - 1762050814
                  - 4150417245
                  - 2154129355
                  - 426522225
                  - 1852507879
                  - 4275313526
                  - 2312317920
                  - 282753626
                  - 1742555852
                  - 4189708143
                  - 2394877945
                  - 397917763
                  - 1622183637
                  - 3604390888
                  - 2714866558
                  - 953729732
                  - 1340076626
                  - 3518719985
                  - 2797360999
                  - 1068828381
                  - 1219638859
                  - 3624741850
                  - 2936675148
                  - 906185462
                  - 1090812512
                  - 3747672003
                  - 2825379669
                  - 829329135
                  - 1181335161
                  - 3412177804
                  - 3160834842
                  - 628085408
                  - 1382605366
                  - 3423369109
                  - 3138078467
                  - 570562233
                  - 1426400815
                  - 3317316542
                  - 2998733608
                  - 733239954
                  - 1555261956
                  - 3268935591
                  - 3050360625
                  - 752459403
                  - 1541320221
                  - 2607071920
                  - 3965973030
                  - 1969922972
                  - 40735498
                  - 2617837225
                  - 3943577151
                  - 1913087877
                  - 83908371
                  - 2512341634
                  - 3803740692
                  - 2075208622
                  - 213261112
                  - 2463272603
                  - 3855990285
                  - 2094854071
                  - 198958881
                  - 2262029012
                  - 4057260610
                  - 1759359992
                  - 534414190
                  - 2176718541
                  - 4139329115
                  - 1873836001
                  - 414664567
                  - 2282248934
                  - 4279200368
                  - 1711684554
                  - 285281116
                  - 2405801727
                  - 4167216745
                  - 1634467795
                  - 376229701
                  - 2685067896
                  - 3608007406
                  - 1308918612
                  - 956543938
                  - 2808555105
                  - 3495958263
                  - 1231636301
                  - 1047427035
                  - 2932959818
                  - 3654703836
                  - 1088359270
                  - 936918000
                  - 2847714899
                  - 3736837829
                  - 1202900863
                  - 817233897
                  - 3183342108
                  - 3401237130
                  - 1404277552
                  - 615818150
                  - 3134207493
                  - 3453421203
                  - 1423857449
                  - 601450431
                  - 3009837614
                  - 3294710456
                  - 1567103746
                  - 711928724
                  - 3020668471
                  - 3272380065
                  - 1510334235
                  - 755167117
                lib:
                  constants:
                    OPENSSL_VERSION_NUMBER: 269488255
                    SSL_OP_ALL: 2147485780
                    SSL_OP_ALLOW_UNSAFE_LEGACY_RENEGOTIATION: 262144
                    SSL_OP_CIPHER_SERVER_PREFERENCE: 4194304
                    SSL_OP_CISCO_ANYCONNECT: 32768
                    SSL_OP_COOKIE_EXCHANGE: 8192
                    SSL_OP_CRYPTOPRO_TLSEXT_BUG: 2147483648
                    SSL_OP_DONT_INSERT_EMPTY_FRAGMENTS: 2048
                    SSL_OP_EPHEMERAL_RSA: 0
                    SSL_OP_LEGACY_SERVER_CONNECT: 4
                    SSL_OP_MICROSOFT_BIG_SSLV3_BUFFER: 0
                    SSL_OP_MICROSOFT_SESS_ID_BUG: 0
                    SSL_OP_MSIE_SSLV2_RSA_PADDING: 0
                    SSL_OP_NETSCAPE_CA_DN_BUG: 0
                    SSL_OP_NETSCAPE_CHALLENGE_BUG: 0
                    SSL_OP_NETSCAPE_DEMO_CIPHER_CHANGE_BUG: 0
                    SSL_OP_NETSCAPE_REUSE_CIPHER_CHANGE_BUG: 0
                    SSL_OP_NO_COMPRESSION: 131072
                    SSL_OP_NO_QUERY_MTU: 4096
                    SSL_OP_NO_SESSION_RESUMPTION_ON_RENEGOTIATION: 65536
                    SSL_OP_NO_SSLv2: 0
                    SSL_OP_NO_SSLv3: 33554432
                    SSL_OP_NO_TICKET: 16384
                    SSL_OP_NO_TLSv1: 67108864
                    SSL_OP_NO_TLSv1_1: 268435456
                    SSL_OP_NO_TLSv1_2: 134217728
                    SSL_OP_PKCS1_CHECK_1: 0
                    SSL_OP_PKCS1_CHECK_2: 0
                    SSL_OP_SINGLE_DH_USE: 0
                    SSL_OP_SINGLE_ECDH_USE: 0
                    SSL_OP_SSLEAY_080_CLIENT_DH_BUG: 0
                    SSL_OP_SSLREF2_REUSE_CERT_TYPE_BUG: 0
                    SSL_OP_TLS_BLOCK_PADDING_BUG: 0
                    SSL_OP_TLS_D5_BUG: 0
                    SSL_OP_TLS_ROLLBACK_BUG: 8388608
                    ENGINE_METHOD_RSA: 1
                    ENGINE_METHOD_DSA: 2
                    ENGINE_METHOD_DH: 4
                    ENGINE_METHOD_RAND: 8
                    ENGINE_METHOD_EC: 2048
                    ENGINE_METHOD_CIPHERS: 64
                    ENGINE_METHOD_DIGESTS: 128
                    ENGINE_METHOD_PKEY_METHS: 512
                    ENGINE_METHOD_PKEY_ASN1_METHS: 1024
                    ENGINE_METHOD_ALL: 65535
                    ENGINE_METHOD_NONE: 0
                    DH_CHECK_P_NOT_SAFE_PRIME: 2
                    DH_CHECK_P_NOT_PRIME: 1
                    DH_UNABLE_TO_CHECK_GENERATOR: 4
                    DH_NOT_SUITABLE_GENERATOR: 8
                    ALPN_ENABLED: 1
                    RSA_PKCS1_PADDING: 1
                    RSA_SSLV23_PADDING: 2
                    RSA_NO_PADDING: 3
                    RSA_PKCS1_OAEP_PADDING: 4
                    RSA_X931_PADDING: 5
                    RSA_PKCS1_PSS_PADDING: 6
                    RSA_PSS_SALTLEN_DIGEST: -1
                    RSA_PSS_SALTLEN_MAX_SIGN: -2
                    RSA_PSS_SALTLEN_AUTO: -2
                    defaultCoreCipherList: >-
                      TLS_AES_256_GCM_SHA384:TLS_CHACHA20_POLY1305_SHA256:TLS_AES_128_GCM_SHA256:ECDHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-AES256-GCM-SHA384:DHE-RSA-AES128-GCM-SHA256:ECDHE-RSA-AES128-SHA256:DHE-RSA-AES128-SHA256:ECDHE-RSA-AES256-SHA384:DHE-RSA-AES256-SHA384:ECDHE-RSA-AES256-SHA256:DHE-RSA-AES256-SHA256:HIGH:!aNULL:!eNULL:!EXPORT:!DES:!RC4:!MD5:!PSK:!SRP:!CAMELLIA
                    TLS1_VERSION: 769
                    TLS1_1_VERSION: 770
                    TLS1_2_VERSION: 771
                    TLS1_3_VERSION: 772
                    POINT_CONVERSION_COMPRESSED: 2
                    POINT_CONVERSION_UNCOMPRESSED: 4
                    POINT_CONVERSION_HYBRID: 6
              abort: {}
              uuid: {}
              ARN: {}
              defaultProfile: default
              configOptInEnv: AWS_SDK_LOAD_CONFIG
              sharedCredentialsFileEnv: AWS_SHARED_CREDENTIALS_FILE
              sharedConfigFileEnv: AWS_CONFIG_FILE
              imdsDisabledEnv: AWS_EC2_METADATA_DISABLED
              domain:
                _stack: []
                active: null
              url: {}
              querystring: {}
              realClock: {}
              clientSideMonitoring: {}
              iniLoader:
                resolvedProfiles: {}
            VERSION: 2.820.0
            Signers: {}
            Protocol:
              Json: {}
              Query: {}
              Rest: {}
              RestJson: {}
              RestXml: {}
            XML: {}
            JSON: {}
            Model: {}
            config:
              credentialProvider:
                providers:
                  - null
                  - null
                  - null
                  - null
                  - null
                  - null
                  - null
                resolveCallbacks: []
              apiVersions: {}
              httpOptions:
                timeout: 120000
              maxRedirects: 10
              paramValidation: true
              sslEnabled: true
              s3ForcePathStyle: false
              s3BucketEndpoint: false
              s3DisableBodySigning: true
              s3UsEast1RegionalEndpoint: legacy
              computeChecksums: true
              convertResponseTypes: true
              correctClockSkew: false
              dynamoDbCrc32: true
              systemClockOffset: 0
              signatureCache: true
              retryDelayOptions: {}
              useAccelerateEndpoint: false
              clientSideMonitoring: false
              endpointCacheSize: 1000
              hostPrefixEnabled: true
              stsRegionalEndpoints: legacy
              credentials: null
              logger: null
              apiVersion: null
              customUserAgent: null
              signatureVersion: null
            EventListeners:
              Core:
                _events:
                  validate:
                    - null
                    - null
                    - null
                    - null
                  afterBuild:
                    - null
                    - null
                    - null
                  restart:
                    - null
                  sign:
                    - null
                    - null
                  validateResponse:
                    - null
                  send:
                    - null
                  httpHeaders:
                    - null
                  httpData:
                    - null
                  httpDone:
                    - null
                  retry:
                    - null
                    - null
                    - null
                    - null
                    - null
                    - null
                  afterRetry:
                    - null
              CorePost:
                _events:
                  extractData:
                    - null
                  extractError:
                    - null
                  httpError:
                    - null
              Logger:
                _events:
                  complete:
                    - null
              Json:
                _events:
                  build:
                    - null
                  extractData:
                    - null
                  extractError:
                    - null
              Rest:
                _events:
                  build:
                    - null
                  extractData:
                    - null
                  extractError:
                    - null
              RestJson:
                _events:
                  build:
                    - null
                  extractData:
                    - null
                  extractError:
                    - null
              RestXml:
                _events:
                  build:
                    - null
                  extractData:
                    - null
                  extractError:
                    - null
              Query:
                _events:
                  build:
                    - null
                  extractData:
                    - null
                  extractError:
                    - null
            events:
              _events: {}
            endpointCache:
              maxSize: 1000
              cache:
                nodeMap: {}
                size: 0
                sizeLimit: 1000
          hooks: {}
          requestCache: {}
          requestQueue:
            options: {}
            pendingPromises: 0
            maxPendingPromises: 2
            maxQueuedPromises: null
            queue: []
          cachedCredentials: null
      version: 2.17.0
      yamlParser:
        serverless:
          $ref: '$["custom"]["serviceReference"]["serverless"]'
      utils:
        serverless:
          $ref: '$["custom"]["serviceReference"]["serverless"]'
      service:
        $ref: '$["custom"]["serviceReference"]'
      variables:
        serverless:
          $ref: '$["custom"]["serviceReference"]["serverless"]'
        service:
          $ref: $
        tracker:
          promiseList: []
          promiseMap: {}
          startTime: 1609886203030
          reported: false
          interval:
            _idleTimeout: -1
            _idlePrev: null
            _idleNext: null
            _idleStart: 5791
            _onTimeout: null
            _repeat: 15000
            _destroyed: true
        deep: []

        deepRefSyntax: {}
        overwriteSyntax: {}
        fileRefSyntax: {}
        slsRefSyntax: {}
        envRefSyntax: {}
        optRefSyntax: {}
        selfRefSyntax: {}
        stringRefSyntax: {}
AssertionError: expected promise to be rejected but it was fulfilled with { Object (serverless, stdoutData, ...) }
        boolRefSyntax: {}
        intRefSyntax: {}
        s3RefSyntax: {}
        cfRefSyntax: {}
        ssmRefSyntax: {}
        strToBoolRefSyntax: {}
        variableResolvers:
          - regex:
              $ref: >-
                $["custom"]["serviceReference"]["serverless"]["variables"]["slsRefSyntax"]
          - regex:
              $ref: >-
                $["custom"]["serviceReference"]["serverless"]["variables"]["envRefSyntax"]
          - regex:
              $ref: >-
                $["custom"]["serviceReference"]["serverless"]["variables"]["optRefSyntax"]
          - regex:
              $ref: >-
                $["custom"]["serviceReference"]["serverless"]["variables"]["selfRefSyntax"]
          - regex:
              $ref: >-
                $["custom"]["serviceReference"]["serverless"]["variables"]["fileRefSyntax"]
          - regex:
              $ref: >-
                $["custom"]["serviceReference"]["serverless"]["variables"]["cfRefSyntax"]
            isDisabledAtPrepopulation: true
            serviceName: CloudFormation
          - regex:
              $ref: >-
                $["custom"]["serviceReference"]["serverless"]["variables"]["s3RefSyntax"]
            isDisabledAtPrepopulation: true
            serviceName: S3
          - regex:
              $ref: >-
                $["custom"]["serviceReference"]["serverless"]["variables"]["stringRefSyntax"]
          - regex:
              $ref: >-
                $["custom"]["serviceReference"]["serverless"]["variables"]["boolRefSyntax"]
          - regex:
              $ref: >-
                $["custom"]["serviceReference"]["serverless"]["variables"]["intRefSyntax"]
          - regex:
              $ref: >-
                $["custom"]["serviceReference"]["serverless"]["variables"]["ssmRefSyntax"]
            isDisabledAtPrepopulation: true
            serviceName: SSM
          - regex:
              $ref: >-
                $["custom"]["serviceReference"]["serverless"]["variables"]["strToBoolRefSyntax"]
          - regex:
              $ref: >-
                $["custom"]["serviceReference"]["serverless"]["variables"]["deepRefSyntax"]
          - regex: {}
            isDisabledAtPrepopulation: true
            serviceName: Serverless Parameters
          - regex: {}
            isDisabledAtPrepopulation: true
            serviceName: Serverless Secrets
          - regex: {}
            isDisabledAtPrepopulation: true
            serviceName: Serverless Outputs
          - regex: {}
            isDisabledAtPrepopulation: true
            serviceName: Serverless Outputs
        options:
          $ref: >-
            $["custom"]["serviceReference"]["serverless"]["providers"]["aws"]["options"]
        variableSyntax: {}
      pluginManager:
        serverless:
          $ref: '$["custom"]["serviceReference"]["serverless"]'
        serverlessConfigFile:
          $ref: $
        cliOptions:
          $ref: >-
            $["custom"]["serviceReference"]["serverless"]["providers"]["aws"]["options"]
        cliCommands:
          - print
        pluginIndependentCommands: {}
        plugins:
          - serverless:
              $ref: '$["custom"]["serviceReference"]["serverless"]'
            commands:
              interactiveCli:
                isHidden: true
                options:
                  app:
                    usage: Dashboard app
                  org:
                    usage: Dashboard org
                lifecycleEvents:
                  - initializeService
                  - setupAws
                  - autoUpdate
                  - tabCompletion
                  - end
            hooks: {}
          - serverless:
              $ref: '$["custom"]["serviceReference"]["serverless"]'
            options:
              $ref: >-
                $["custom"]["serviceReference"]["serverless"]["providers"]["aws"]["options"]
            commands:
              config:
                usage: Configure Serverless
                configDependent: false
                lifecycleEvents:
                  - config
                options:
                  autoupdate:
                    usage: >-
                      Turn on auto update mechanism (turn off via
                      "--no-autoupdate")
                commands:
                  credentials:
                    usage: >-
                      Configures a new provider profile for the Serverless
                      Framework
                    lifecycleEvents:
                      - config
                    validProviders: {}
                    options:
                      provider:
                        usage: 'Name of the provider. Supported providers: "aws"'
                        required: true
                        shortcut: p
                  tabcompletion:
                    usage: Configure command line <tab> completion
                    commands:
                      install:
                        usage: Install a <tab> completion for chosen shell
                        lifecycleEvents:
                          - install
                        options:
                          shell:
                            usage: >-
                              Shell for which <tab> completion should be
                              installed. Supported options: bash (default), zsh,
                              fish 
                            shortcut: s
                          location:
                            usage: Custom location for shell config
                            shortcut: l
                      uninstall:
                        usage: Uninstall a <tab> completion for chosen shell
                        lifecycleEvents:
                          - uninstall
            hooks: {}
          - serverless:
              $ref: '$["custom"]["serviceReference"]["serverless"]'
            options:
              $ref: >-
                $["custom"]["serviceReference"]["serverless"]["providers"]["aws"]["options"]
            commands:
              create:
                usage: Create new Serverless service
                lifecycleEvents:
                  - create
                options:
                  template:
                    usage: |-
                      Template for the service. Available templates: 
                                                                   "aws-clojure-gradle", "aws-clojurescript-gradle", "aws-nodejs", "aws-nodejs-typescript", "aws-alexa-typescript", "aws-nodejs-ecma-script", "aws-python", "aws-python3"
                                                                   "aws-groovy-gradle", "aws-java-maven", "aws-java-gradle", "aws-kotlin-jvm-maven", "aws-kotlin-jvm-gradle", "aws-kotlin-jvm-gradle-kts", "aws-kotlin-nodejs-gradle", "aws-scala-sbt"
                                                                   "aws-csharp", "aws-fsharp", "aws-go", "aws-go-dep", "aws-go-mod", "aws-ruby", "aws-provided"
                                                                   "tencent-go", "tencent-nodejs", "tencent-python", "tencent-php"
                                                                   "azure-csharp", "azure-nodejs", "azure-nodejs-typescript", "azure-python"
                                                                   "cloudflare-workers", "cloudflare-workers-enterprise", "cloudflare-workers-rust"
                                                                   "fn-nodejs", "fn-go"
                                                                   "google-nodejs", "google-python", "google-go"
                                                                   "kubeless-python", "kubeless-nodejs"
                                                                   "knative-docker"
                                                                   "openwhisk-java-maven", "openwhisk-nodejs", "openwhisk-php", "openwhisk-python", "openwhisk-ruby", "openwhisk-swift"
                                                                   "spotinst-nodejs", "spotinst-python", "spotinst-ruby", "spotinst-java8"
                                                                   "twilio-nodejs"
                                                                   "aliyun-nodejs"
                                                                   "plugin"
                                                                   "hello-world"
                    shortcut: t
                  template-url:
                    usage: 'Template URL for the service. Supports: GitHub, BitBucket'
                    shortcut: u
                  template-path:
                    usage: Template local path for the service.
                  path:
                    usage: >-
                      The path where the service should be created (e.g. --path
                      my-service)
                    shortcut: p
                  name:
                    usage: >-
                      Name for the service. Overwrites the default name of the
                      created service.
                    shortcut: 'n'
            hooks: {}
          - serverless:
              $ref: '$["custom"]["serviceReference"]["serverless"]'
            options:
              $ref: >-
                $["custom"]["serviceReference"]["serverless"]["providers"]["aws"]["options"]
            commands:
              install:
                usage: >-
                  Install a Serverless service from GitHub or a plugin from the
                  Serverless registry
                lifecycleEvents:
                  - install
                options:
                  url:
                    usage: URL of the Serverless service on GitHub
                    required: true
                    shortcut: u
                  name:
                    usage: Name for the service
                    shortcut: 'n'
            hooks: {}
          - serverless:
              $ref: '$["custom"]["serviceReference"]["serverless"]'
            options:
              $ref: >-
                $["custom"]["serviceReference"]["serverless"]["providers"]["aws"]["options"]
            servicePath: /tmp/tmpdirs-serverless/45eb/85c788
            packagePath: /tmp/tmpdirs-serverless/45eb/85c788/.serverless
            defaultExcludes:
              - .git/**
              - .gitignore
              - .DS_Store
              - npm-debug.log
              - yarn-*.log
              - .serverless/**
              - .serverless_plugins/**
            commands:
              package:
                usage: Packages a Serverless service
                configDependent: true
                lifecycleEvents:
                  - cleanup
                  - initialize
                  - setupProviderConfiguration
                  - createDeploymentArtifacts
                  - compileLayers
                  - compileFunctions
                  - compileEvents
                  - finalize
                options:
                  stage:
                    usage: Stage of the service
                    shortcut: s
                  region:
                    usage: Region of the service
                    shortcut: r
                  package:
                    usage: Output path for the package
                    shortcut: p
                  app:
                    usage: Dashboard app
                  org:
                    usage: Dashboard org
                commands:
                  function:
                    type: entrypoint
                    lifecycleEvents:
                      - package
            hooks: {}
          - serverless:
              $ref: '$["custom"]["serviceReference"]["serverless"]'
            options:
              $ref: >-
                $["custom"]["serviceReference"]["serverless"]["providers"]["aws"]["options"]
            commands:
              deploy:
                usage: Deploy a Serverless service
```
</details>

Reply:
It looks as quirk in how `print` command resolves confing. I would investigate what's `svc` here: https://github.com/serverless/serverless/blob/44603830a6c004672da5f8544ba272eed57cdca7/lib/plugins/print.js#L99 (it should be plain config as is at `serverless.yml` and check further down the path where it's lost

Reply:
Sure thing - but any chance of doing this in a followup PR instead of blocking this one? I don't mind having a go at converting the tests to the new method, but these bugs are making it more complex and I think would benefit from being tackled on it's own.

Reply:
<details>
<summary>
Seems that `svc` does have `unresolvedVariablesNotificationMode` - will follow the code and try to find where it disappears
</summary>

```
2021-01-06T18:52:39.808Z ℹ serverless:test:stdout <ref *2> {
  service: { name: 'test-variables-9qs3701' },
  provider: {
    name: 'aws',
    runtime: 'nodejs12.x',
    stage: 'dev',
    variableSyntax: '\\${([^{}:]+?(?:\\(|:)(?:[^:{}][^{}]*?)?)}',
    region: 'us-east-1',
    versionFunctions: true
  },
  custom: <ref *1> {
    importedFile: { foo: 'bar' },
    importedFileWithKey: 'bar',
    awsVariable: '${AWS::Region}',
    cloudFormationReference: '${AnotherResource}',
    selfReference: 'bar',
    serviceReference: Service {
      serverless: [Serverless],
      service: 'test-variables-9qs3701',
      serviceObject: [Object],
      provider: [Object],
      custom: [Circular *1],
      pluginsData: {},
      functions: {},
      package: {},
      configValidationMode: 'warn',
      serviceFilename: 'serverless.yml',
      unresolvedVariablesNotificationMode: 'error',
      layers: {},
      initialServerlessConfig: [Circular *2],
      isDashboardMonitoringPreconfigured: false,
      plugins: undefined,
      resources: undefined,
      disabledDeprecations: undefined,
      app: undefined,
      org: undefined,
      outputs: undefined
    },
    nestedRef: 'nestedVal',
    nestedVal: { prop: 'resolvedNested' },
    nestedReference: 'resolvedNested',
    myVariable: undefined
  },
  unresolvedVariablesNotificationMode: 'error'
}
```
</details>

Reply:
Technically we're not meant to test `print` out in this `variables.test.js`

But when adding new property ideally it should be  onfirmed that it works as intended in all areas, otherwise submission is not really complete (and we avoid introducing incomplete/WIP merges to `master`)

Reply:
> Technically we're not meant to test print out in this variables.test.js

I think the bug is elsewhere, as it looks like variables have already been resolved by the time `print` is being called (e.g. `custom.myVariable` is `undefined` and the warning has been printed) - tbh it could be the test runner is just truncating the output.

Reply:
Seems that the property is available on `service` (I'm guessing because I'm assigning it in `Service.js`) - does that sound like a valid way to access it, or should I continue to try and get it via `serverless.config`?

Reply:
> Seems that the property is available on `service`

On `service` instance? Or on `service` config property (which is accessible at `service.service`) ?

Reply:
Yes, [on the `this.service](https://github.com/serverless/serverless/pull/8710/files#diff-9c8d8c1e18a753e5de9260e0325e7dacfc541a932945b2701be43eab18b3e953R995) instance.

I've got things working now, and so have written tests + cleaned up my commit - should be ready for re-review :)